python_modules/libraries/dagster-dg/dagster_dg/config.py

```diff
@@ -52,115 +168,152 @@ class DgConfig:
+
+    @classmethod
+    def from_raw(cls, raw: "DgRawProjectConfig") -> Self:
+        return cls(
+            root_module=raw["root_module"],
+            components_module=raw.get("components_module", DgProjectConfig.components_module),
+        )
+
+
+class DgRawProjectConfig(TypedDict):
+    root_module: Required[str]
+    components_module: NotRequired[str]
+
+
+# ########################
+# ##### WORKSPACE
+# ########################
+
+
+@dataclass
+class DgWorkspaceConfig:
+    pass
+
+    @classmethod
+    def from_raw(cls, raw: "DgRawWorkspaceConfig") -> Self:
+        return cls()
+
+
+class DgRawWorkspaceConfig(TypedDict):
+    pass
+
+
+# ########################
+# ##### CLI CONFIG
 # ########################
 
 
 def normalize_cli_config(
     cli_options: Mapping[str, object], cli_context: click.Context
-) -> DgPartialConfig:
+) -> DgRawCliConfig:
     # Remove any options that weren't explicitly provided.
     filtered_options = {
         key: value
         for key, value in cli_options.items()
         if cli_context.get_parameter_source(key) != ParameterSource.DEFAULT
     }
-    return _normalize_dg_partial_config(filtered_options)
+    return _validate_cli_config(filtered_options)
 
 
-_CLI_CONTEXT_CONFIG_KEY = "config"
+def _validate_cli_config(cli_opts: Mapping[str, object]) -> DgRawCliConfig:
+    try:
+        for key, type_ in DgRawCliConfig.__annotations__.items():
+            _validate_cli_config_setting(cli_opts, key, type_)
+        _validate_cli_config_no_extraneous_keys(cli_opts)
+    except DgValidationError as e:
+        _raise_cli_config_validation_error(str(e))
+    return cast(DgRawCliConfig, cli_opts)
 
 
-def set_config_on_cli_context(cli_context: click.Context, config: DgPartialConfig) -> None:
-    cli_context.ensure_object(dict)
-    cli_context.obj[_CLI_CONTEXT_CONFIG_KEY] = config
+def _validate_cli_config_setting(cli_opts: Mapping[str, object], key: str, type_: type) -> None:
+    if key in cli_opts and not isinstance(cli_opts[key], type_):
+        raise DgValidationError(f"`{key}` must be a {type_.__name__}.")
 
 
-def has_config_on_cli_context(cli_context: click.Context) -> bool:
-    return _CLI_CONTEXT_CONFIG_KEY in cli_context.ensure_object(dict)
+def _validate_cli_config_no_extraneous_keys(cli_opts: Mapping[str, object]) -> None:
+    extraneous_keys = [k for k in cli_opts.keys() if k not in DgRawCliConfig.__annotations__]
+    if extraneous_keys:
+        raise DgValidationError(f"Unrecognized fields: {extraneous_keys}")
 
 
-def get_config_from_cli_context(cli_context: click.Context) -> DgPartialConfig:
-    cli_context.ensure_object(dict)
-    return cli_context.obj[_CLI_CONTEXT_CONFIG_KEY]
+def _raise_cli_config_validation_error(message: str) -> None:
+    raise DgError(f"Error in CLI options: {message}")
 
 
 # ########################
-# ##### CONFIG FILE LOADING
+# ##### FILE CONFIG
 # ########################
 
+DgFileConfigDirectoryType = Literal["workspace", "project"]
```

Comment:
Is this preferable to an enum?

Reply:
I think so because it can be used to type the raw string coming from the config file.